src/include/daos_obj.h

```diff
@@ -0,0 +1,786 @@
+ * array. \a rx_idx is the first array index of the extent and \a rx_nr is the
+ * number of records covered by the extent.
+ */
+typedef struct daos_recx {
+	/** Indice of the first record in the extent */
+	uint64_t	rx_idx;
+	/**
+	 * Number of contiguous records in the extent
+	 * If \a rx_nr is equal to 1, the extent is composed of a single record
+	 * at indice \a rx_idx
+	 */
+	uint64_t	rx_nr;
+} daos_recx_t;
+
+/** Type of the value accessed in an IOD */
+typedef enum {
+	/** is a dkey */
+	DAOS_IOD_NONE		= 0,
+	/** one indivisble value udpate atomically */
+	DAOS_IOD_SINGLE		= 1,
+	/** an array of records where each record is update atomically */
+	DAOS_IOD_ARRAY		= 2,
+} daos_iod_type_t;
+
+/**
+ * An I/O descriptor is a list of extents (effectively records associated with
+ * contiguous array indices) to update/fetch in a particular array identified by
+ * its akey.
+ */
+typedef struct daos_io_desc {
+	/** akey for this iod */
+	daos_key_t		iod_name;
+	/** akey checksum */
+	daos_csum_buf_t		iod_kcsum;
+	/*
+	 * Type of the value in an iod can be either a single type that is
+	 * always overwritten when updated, or it can be an array of EQUAL sized
+	 * records where the record is updated atomically. Note that an akey can
+	 * only support one type of value which is set on the first update. If
+	 * user attempts to mix types in the same akey, the behavior is
+	 * undefined, even after the object, key, or value is punched. If
+	 * \a iod_type == DAOS_IOD_SINGLE, then iod_nr has to be 1, and
+	 * \a iod_size would be the size of the single atomic value. The idx is
+	 * ignored and the rx_nr is also required to be 1.
+	 */
+	daos_iod_type_t		iod_type;
+	/** Size of the single value or the record size of the array */
+	daos_size_t		iod_size;
+	/*
+	 * Number of entries in the \a iod_recxs, \a iod_csums, and \a iod_eprs
+	 * arrays, should be 1 if single value.
+	 */
+	unsigned int		iod_nr;
+	/*
+	 * Array of extents, where each extent defines the index of the first
+	 * record in the extent and the number of records to access. If the
+	 * type of the iod is single, this is ignored.
+	 */
+	daos_recx_t		*iod_recxs;
+	/*
+	 * Checksum associated with each extent. If the type of the iod is
+	 * single, will only have a single checksum.
+	 */
+	daos_csum_buf_t		*iod_csums;
+	/** Epoch range associated with each extent */
+	daos_epoch_range_t	*iod_eprs;
+} daos_iod_t;
+
+/**
+ * A I/O map represents the physical extent mapping inside an array for a
+ * given range of indices.
+ */
+typedef struct daos_io_map {
+	/** type of akey value (SV or AR)*/
+	daos_iod_type_t		 iom_type;
+	/**
+	 * Number of extents in the mapping, that's the size of all the
+	 * external arrays listed below. 1 for SV.
+	 */
+	unsigned int		 iom_nr;
+	/** Size of the single value or the record size */
+	daos_size_t		 iom_size;
+	/**
+	 * The recx with the lowest offset within the requested extents
+	 * daos_io_desc::iod_recxs
+	 */
+	struct daos_recx	 iom_recx_lo;
```

Comment:
why are these needed?   if iom_recxs is simply sorted, can't the user just infer the lo and the hi ?

Reply:
we don't want to fetch all recxs in some situations, and just the hi / lo

Reply:
indeed, require user to always allocate array is not very nice, specially when most of them don't really care about those recxs